Add mixed-layer depth map_blocks example (#43)
NCAR/esds: posts/map_blocks_example.md

---
author: Matt Long
date: 2021-4-22
tags: xarray, dask, pop
---

# How to Use `xarray.map_blocks` within a Mixed-Layer Depth Calculation

Within this example, we cover how to use `xarray.map_blocks` to calculate the mixed-layer depth within CESM POP model output.

> This calculation is "embarassingly parellel" such that each calculation is done within a single a column. The calculation ***should*** be easily computed within each column across the model domain. This is where `map_blocks` can be used to improve the performance of this metric.

Reviewer: parellel" --> parallel,"

## Imports and Data Ingestion

We use the following libraries/packages within this example
```python
import os

from scipy import interpolate
import numpy as np
import xarray as xr

import matplotlib.pyplot as plt

import pop_tools
```

Next, we read in the data using `xarray` and subset for temperature and salinity, subsetting for a smaller portion of the domain for testing purposes

> 

Reviewer: This post might be more powerful if we applied to the full dataset, demonstrating the scaling, etc.

Also, we might consider retaining/curating datasets like this for testing/demo purposes. If we have a collection of datasets like this, we can provide people with working examples. 

I noticed that these particular datasets are unwieldy because somewhere along the way in their generation, `xarray` put the `member_id` and perhaps the `time` dimension into the grid variables (i.e., `TAREA`, `KMT`, etc.). Super annoying.

Author: I think keeping it shorter for this post might be okay with the subset - in terms of executable notebooks, it might be helpful to contribute to the [NCAR Pangeo Gallery](https://gallery.pangeo.io/repos/NCAR/notebook-gallery/notebooks/Run-Anywhere/Ocean-Heat-Content/OHC_tutorial.html)? That way we can help increase the presence there instead of having our own separate gallery.

```
ds = xr.merge((
    xr.open_dataset('/glade/p/cgd/oce/projects/cesm2-marbl/xpersist_cache/3d_fields/TEMP-presentday-monclim.nc'),
    xr.open_dataset('/glade/p/cgd/oce/projects/cesm2-marbl/xpersist_cache/3d_fields/SALT-presentday-monclim.nc'),    
))

ds = ds[['TEMP', 'SALT']]

# subset for testing purposes
ds = ds.isel(member_id=slice(0, 2), nlat=slice(2, 5), nlon=slice(1, 5)) 
```

## Data Operation

> Next, we setup a function to compute mixed layer depth using a difference in density (sigma)

Reviewer: Change to:
> Next, we setup a function to compute the mixed layer depth (MLD). The MLD is defined as the point in the water column where the potential density (`sigma`) exceeds the surface density by a specified threshold (`dsigma`).

```python
def mld_dsigma(SALT, TEMP, dsigma=0.03, rho_chunks={'nlat': 16, 'nlon': 16}):
    """
    Compute MLD based on ∆σ criterion. Uses xarray.map_blocks.
    
    Parameters
    ----------
    
    SALT : xarray.DataArray
      Salinity
    TEMP : xarray.DataArray
      Potential temperature
    dsigma : float, optional
      The value for ∆σ.

>       

Reviewer: add `rho_chunks` paramter

    MLD : xarray.DataArray
      The MLD (m) defined as the point in the water column where
      density exceeds rho[0] + dsigma.      
    """
    
    # determine dimensionality
    dims_in = SALT.dims
    assert dims_in == TEMP.dims, 'dimension mismatch'
    assert 'z_t' in SALT.coords, 'z_t not found in SALT coords'

    # drop ancillary coordinates (this may not be necessary)
    SALT = SALT.reset_coords(drop=True)
    TEMP = TEMP.reset_coords(drop=True)
    
    # compute density
    rho = pop_tools.eos(SALT.chunk({'z_t': 10}), 
                        TEMP.chunk({'z_t': 10}), 
                        depth=SALT.z_t * 0.).compute()
    
    # add these coordinates which creep in somewhere, I think when xarray does the unstack
    # without these, I get an error: not expecting {'nlat', 'nlon'}
    # chunking here is arbitrary and maybe suitable for the global POP_gx1v7 grid
    if 'nlat' in rho.dims and 'nlon' in rho.dims:
        rho = rho.assign_coords({
            'nlat': xr.DataArray(np.arange(len(SALT.nlat)), dims=('nlat')),
            'nlon': xr.DataArray(np.arange(len(SALT.nlon)), dims=('nlon')),
        })
        
    rho = rho.chunk(rho_chunks).persist()
    

>     # compute and return MLD - this is where the parallelization comes in

Reviewer: Add here
```python
    # compute and return MLD
    template = rho.isel(z_t=0).drop('z_t')
    template.attrs['long_name'] = 'MLD'
    template.attrs['units'] = SALT.z_t.attrs['units']
    template.name = 'MLD'
```

    return xr.map_blocks(
        interp_mld_dsigma, rho,
        kwargs=dict(dsigma=dsigma), 

>         template=rho.isel(z_t=0).drop('z_t'),

Reviewer: change to
```python 
template=template
```

    )
```

### Applying Mixed Layer Depth Calculation
This function, `interp_mld_dsigma` is designed to be called within `xr.map_blocks`. It assumes that `rho_in` has a depth dimension, `z_t` and some number of unspecified other dimensions, `non_vertical_dims`.

> The non_vertical_dims are "stacked" using `xarray.stack` and the code loops over these dimensions, performing linear interpolation in `z_t` at each location.

Reviewer: use `non_vertical_dims`

Author: What do you mean by this?

Reviewer: > The `non_vertical_dims` are "stacked" using...

> The data are return unstacked with the dimensions ordered as the arrived in `rho_in`.

Reviewer: "are return[ed]" 

"as [they] arrived..."

We define the function here, then immediately call it to test it; arbitrarily, we're calling it on TEMP rather than density.

```python
def interp_mld_dsigma(rho_in, dsigma=0.03):
    """compute MLD at point using interpolation"""
    

>     # TODO: this will probably fail if non_vertical_dims = []

Reviewer: Worth addressing?

    #       In that case, we might use `expand_dims` to add a dummy dimension
    non_vertical_dims = [d for d in rho_in.dims if d not in ['z_t']] 
    rho_stack = rho_in.stack(non_vertical_dims=non_vertical_dims)
    mld_stack = xr.full_like(rho_stack.isel(z_t=slice(0, 1)), fill_value=np.nan)
    N = len(rho_stack.non_vertical_dims)

>     z_t = rho_in.z_t * 1e-2

Reviewer: remove the `* 1e-2`

This converts from `cm` to `m`, but above we set the units to `SALT.z_t.attrs['units']`. I think it's more straightforward to leave the units as they are.

---
author: Matt Long
date: 2021-4-22
tags: xarray, dask, pop
---

> # How to Use `xarray.map_blocks` for Vertical Interpolation of a 3D Field

Reviewer 2: ```suggestion
# How to Use `xarray.map_blocks` to easily parallelize a function that expects xarray objects
```

`map_blocks` is nice and convenient but inefficient.

`map_blocks` takes a block of each dask array, wraps it up in xarray metadata, then wraps all that in to a dataset and sends it to your function. There is a task in the graph for each step for each block. It's a lot of tasks!

 We shouldn't recommend it as a general solution. For fast 1D interpolation, see https://xarray.pydata.org/en/stable/examples/apply_ufunc_vectorize_1d.html

Reviewer: Should this example be using `apply_ufunc_1d`?

Reviewer 2: That's what I would do since this function is not too bad to write with plain numpy. But the current form is still a useful demonstration.

    xr.open_dataset('/glade/p/cgd/oce/projects/cesm2-marbl/xpersist_cache/3d_fields/SALT-presentday-monclim.nc'),
))

ds = ds[['TEMP', 'SALT']]

# subset for testing purposes
ds = ds.isel(member_id=slice(0, 2), nlat=slice(2, 5), nlon=slice(1, 5))
```

## Data Operation

Next, we setup a function to compute the mixed layer depth (MLD). The MLD is defined as the point in the water column where the potential density (`sigma`) exceeds the surface density by a specified threshold (`dsigma`).

```python
def mld_dsigma(SALT, TEMP, dsigma=0.03, rho_chunks={'nlat': 16, 'nlon': 16}):
    """
    Compute MLD based on ∆σ criterion. Uses xarray.map_blocks.

    Parameters
    ----------

    SALT : xarray.DataArray
      Salinity
    TEMP : xarray.DataArray
      Potential temperature
    dsigma : float, optional
      The value for ∆σ.
    rho_chunks: dictionary, optional
      Dimension chunk parameters to chunk rho by

    Returns
    -------

    MLD : xarray.DataArray
      The MLD (m) defined as the point in the water column where
      density exceeds rho[0] + dsigma.
    """

    # determine dimensionality
    dims_in = SALT.dims
    assert dims_in == TEMP.dims, 'dimension mismatch'

>     assert 'z_t' in SALT.coords, 'z_t not found in SALT coords'

Reviewer 2: or you could use cf-xarray and not assume these names! (yet another possible follow up post)

    # drop ancillary coordinates (this may not be necessary)

>     SALT = SALT.reset_coords(drop=True)

Reviewer 2: Absolutely necessary. `map_blocks` takes a block of each dask array, wraps it up in xarray metadata, then wraps all that in to a dataset and sends it to your function. You should only send the data you need to avoid a lot of unnecessary computations.

    TEMP = TEMP.reset_coords(drop=True)

    # compute density

>     rho = pop_tools.eos(SALT.chunk({'z_t': 10}),
>                         TEMP.chunk({'z_t': 10}),
>                         depth=SALT.z_t * 0.).compute()

Reviewer 2: `z_t` is a "core dimension" for this calculation. You can't avoid having it be single-chunked in this case. Well you can by computing :P but that isn't right...

```suggestion
    assert len(SALT.chunks["z_t"]) == 1
    assert len(TEMP.chunks["z_t"]) == 1
    rho = pop_tools.eos(SALT, TEMP, depth=SALT.z_t * 0.)
```

Reviewer: I am not sure I understand this entirely. Are you referring to a limitation in the `apply_ufunc` call in `eos.py`:
https://github.com/NCAR/pop-tools/blob/9563759ab063ae51356af18eab6cabe901e907bc/pop_tools/eos.py#L116
?

Also, I found that the `.compute` on `rho` seemed to improve performance, perhaps by truncating the task graph. This may be taken care of by the `persist` call below.

Reviewer 2: I think my lowest order question is: why is this function not entirely lazy? Isn't `rho` potentially quite large; in which case the `compute` here would break things.

I think it should be lazy; in which case the only way the MLD calculation can work is if there is a single chunk along `z_t`. Hence my suggestion to ensure that.

Reviewer: If rho is very large the `compute` will break things. In my case, `rho` fit in memory, so this didn't break things. However, I think more testing is needed as I subsequently added the `.persist` call below. I found that at least one of these (`compute` or `persist`) was necessary, otherwise the calculation just bogged down with too many tasks—but it might be that `persist` is all that's needed.

To be clear, for the MLD calculation, all dimensions *but* `z_t` can be chunked. However, if these all have a chunk size of one, there are too many tasks and it doesn't work.

Reviewer 2: > n my case, rho fit in memory, so this didn't break things. 

We should at least describe these tradeoffs. I get the sense that people are looking at this blog as a kind of "best practices" recommendation, so we should be careful to do that. So I would still prefer a fully lazy implementation followed by a discussion of when things might break.

> To be clear, for the MLD calculation, all dimensions but z_t can be chunked. However, if these all have a chunk size of one, there are too many tasks and it doesn't work.

Yes this is what I was trying to say in a not very clear way. The proposed assert statements enforces this requirement on the inputs: `SALT`, `TEMP`. So then it is satisfied on `rho`

The "too many tasks" issue is a byproduct of using `map_blocks`. That's just how it works. It should ultimately benefit from ongoing dask work to make dealing with large graphs easier.

Reviewer: > The "too many tasks" issue is a byproduct of using map_blocks. That's just how it works. It should ultimately benefit from ongoing dask work to make dealing with large graphs easier.

I think I managed to set this up so that the interpolation kernel operates on chunks with `len(non_vertical_dims) > 1`—so the parallelization is not across every `i, j` column (for example), but over blocks of data chunked so as to yield a reasonable number of tasks and appropriate amortization of chunk memory overhead across substantial compute.

    # add these coordinates which creep in somewhere, I think when xarray does the unstack
    # without these, I get an error: not expecting {'nlat', 'nlon'}

>     # chunking here is arbitrary and maybe suitable for the global POP_gx1v7 grid

Reviewer 2: curious to see what the error is.

Reviewer: I can't recall exactly but I think there are instances where `xarray` creates coordinates, using something like `nlat = np.arange(N)` when it finds them missing. So this was necessary for the output to conform with the template.

    if 'nlat' in rho.dims and 'nlon' in rho.dims:
        rho = rho.assign_coords({
            'nlat': xr.DataArray(np.arange(len(SALT.nlat)), dims=('nlat')),
            'nlon': xr.DataArray(np.arange(len(SALT.nlon)), dims=('nlon')),
        })

>     rho = rho.chunk(rho_chunks).persist()

Reviewer 2: I would remove this and the `rho_chunks` kwarg. `rho` will be chunked like `TEMP` and `SALT` if you don't compute them earlier.

Reviewer: `rho` should not be chunked in `z_t` as all `z_t` are required in the kernel interpolation function. Chunk sizes of 1 lead to too many tasks and the computation does not proceed.

Reviewer: I think we can remove the kwarg, but check to ensure that `z_t` is not chunked in `SALT` and `TEMP`.

>     # Setup a template

Reviewer 2: We should explain why this is needed. There is some text here: https://xarray.pydata.org/en/stable/dask.html#map-blocks . We should also link to that doc page.

>     template = rho.isel(z_t=0).drop('z_t')

Reviewer 2: minor comment
```suggestion
    template = rho.isel(z_t=0, drop=True)
```

>     template.attrs['long_name'] = 'MLD'

Reviewer 2: ```suggestion
    # the following attributes describe the output and should be set on the template 
    # when provided to map_blocks
    template.attrs['long_name'] = 'MLD'
```

    template.attrs['units'] = SALT.z_t.attrs['units']
    template.name = 'MLD'

    # compute and return MLD - this is where the parallelization comes in
    return xr.map_blocks(
        interp_mld_dsigma, rho,
        kwargs=dict(dsigma=dsigma),
        template=template,
    )
```

### Applying Mixed Layer Depth Calculation

This function, `interp_mld_dsigma` is designed to be called within `xr.map_blocks`. It assumes that `rho_in` has a depth dimension, `z_t` and some number of unspecified other dimensions, `non_vertical_dims`.

> The non_vertical_dims are "stacked" using `xarray.stack` and the code loops over these dimensions, performing linear interpolation in `z_t` at each location.

Reviewer 2: ```suggestion
The `non_vertical_dims` are "stacked" using `xarray.stack` and the code loops over these dimensions, performing linear interpolation in `z_t` at each location.
```

The data are returned unstacked with the dimensions ordered as they arrived in `rho_in`.

We define the function here, then immediately call it to test it; arbitrarily, we're calling it on TEMP rather than density.

```python
def interp_mld_dsigma(rho_in, dsigma=0.03):
    """compute MLD at point using interpolation"""

>     non_vertical_dims = [d for d in rho_in.dims if d not in ['z_t']]

Reviewer 2: minor comment because I like set ops haha

```suggestion
    non_vertical_dims = set(rho_in.dims) - {"z_t"}
```

Reviewer: I like set ops too! But I don't think we can do it this way because `set` does not preserve order.

    rho_stack = rho_in.stack(non_vertical_dims=non_vertical_dims)

>     mld_stack = xr.full_like(rho_stack.isel(z_t=slice(0, 1)), fill_value=np.nan)

Reviewer 2: ```suggestion
    mld_stack = xr.full_like(rho_stack.isel(z_t=0), fill_value=np.nan)
```

Reviewer: ```suggestion
    mld_stack = xr.full_like(rho_stack.isel(z_t=0, drop=True), fill_value=np.nan)
```

Otherwise, we get
>ValueError: Result from applying user function has unexpected coordinate variables {'z_t'}.

>     N = len(rho_stack.non_vertical_dims)

Reviewer 2: ```suggestion
    N = rho_stack.sizes["non_vertical_dims"]
```

    z_t = rho_in.z_t

    for i in range(N):
        # if all NaN, skip this column

>         if np.isnan(rho_stack[:, i]).all():

Reviewer 2: The power and convenience of `map_blocks` comes from the ability to use xarray operations in your parallelized function. So we should do that here.
```suggestion
        if rho_stack[:, i].isnull().all():
```

A good followup post would be the numpy only `apply_ufunc` version of this, followed by an update with a numba version. So there's a nice progression showing how to quickly iterate and develop functionality with `map_blocks` and then move to `apply_ufunc` for more performance, followed by `numba` for even more performance.

            continue

        # if the whole column has density less than the threshold, set MLD to deepest point
        if (rho_stack[:, i] < rho_stack[0, i] + dsigma).all():
            k = np.where(~np.isnan(rho_stack[:, i]))[0]
            mld_stack[:, i] = z_t[k[-1]]

        # linearly interpolate
        else:
            f = interpolate.interp1d(
                rho_stack[:, i], z_t,
                assume_sorted=False,
            )
            mld_stack[:, i] = f(rho_stack[0, i] + dsigma)

>     return mld_stack.unstack().isel(z_t=0, drop=True).transpose(*non_vertical_dims)

Reviewer 2: ```suggestion
    return mld_stack.unstack().transpose(*non_vertical_dims)
```

Reviewer: This requires further changes in the loop below (all instances of `mld_stack[:, i]` need to be changed `mld_stack[i]`

tags: xarray, dask, pop
---

# How to Use `xarray.map_blocks` for Vertical Interpolation of a 3D Field

Within this example, we cover how to use `xarray.map_blocks` to calculate the mixed-layer depth within the CESM POP model output.

This calculation is "embarassingly parallel" such that each calculation is done within a single a column. The calculation **_should_** be easily computed within each column across the model domain. This is where `map_blocks` can be used to improve the performance of this metric.

## Imports and Data Ingestion

We use the following libraries/packages within this example

```python
import os

from scipy import interpolate
import numpy as np
import xarray as xr

import matplotlib.pyplot as plt

import pop_tools
```

Next, we read in the data using `xarray` and subset for temperature and salinity, subsetting for a smaller portion of the domain for testing purposes

```python
ds = xr.merge((
    xr.open_dataset('/glade/p/cgd/oce/projects/cesm2-marbl/xpersist_cache/3d_fields/TEMP-presentday-monclim.nc', chunks={}),
    xr.open_dataset('/glade/p/cgd/oce/projects/cesm2-marbl/xpersist_cache/3d_fields/SALT-presentday-monclim.nc', chunks={}),
))

> ds = ds[['TEMP', 'SALT']]

Reviewer 3: @mgrover1, not sure if you still want to include the `persist()` step here?

```suggestion
ds = ds[['TEMP', 'SALT']].persist()
```
